LSM: panic when compacting an empty level (index out of range in pickCompactNodes)

## Symptoms

The server can **panic** under heavy write load (e.g. SIFT 1M vectors via batch upsert / benchmarks):

```
panic: runtime error: index out of range [0] with length 0
```

The stack points to `pickCompactNodes` in `internal/storage/tree/tree_compact.go`, where `t.nodes[level][0]` is accessed.

## Root cause (scheduler vs. tree state)

1. `tryTriggerCompact(level)` asynchronously enqueues a "compact level `level`" task on `levelCompactCh` via a goroutine.
2. The same level can receive **multiple** compaction triggers in a short window (especially level 0 under sustained writes).
3. The first `compactLevel(level)` may merge and `removeNodes` so that **all** SSTs at that level are gone and `t.nodes[level]` becomes an **empty** slice.
4. A **second** queued task still runs `compactLevel(level)` for the same level; the code assumes at least one node and indexes `t.nodes[level][0]` → panic.

This is not "the vector index is full"; it is a boundary case where **compaction work is queued while the in-memory LSM view no longer has nodes at that level**. Smaller datasets (e.g. Fashion-MNIST) trigger compaction less often, so the race is rarer; large bulk loads make it more likely.

## Relevant code

Start of `pickCompactNodes` (no empty-level guard today):

```go
startKey := t.nodes[level][0].Start()
endKey := t.nodes[level][0].End()
```

## Suggested fix

- In `pickCompactNodes`: if `level` is invalid or `len(t.nodes[level]) == 0`, return `nil`.
- In `compactLevel`: if `pickedNodes` is empty, log and return early (avoid opening writers / further work).

(Optional follow-up: dedupe or avoid redundant same-level compaction enqueue; correctness can be fixed independently.)

## Reproduction notes

Sustained heavy upserts so level 0 keeps crossing compaction thresholds; the panic is timing-dependent, not guaranteed on every run.

---

*Observed while running local benchmarks. Distinct from "invalid sstable" corruption (crash, full disk, etc.).*

--- a/internal/storage/tree/tree_compact.go
+++ b/internal/storage/tree/tree_compact.go
@@ -39,6 +39,10 @@ func (t *LSMTree) compactLevel(level int) {
 
 	// get nodes in level i, and compact them to level i + 1
 	pickedNodes := t.pickCompactNodes(level)
+	if len(pickedNodes) == 0 {
+		logger.Debug("No nodes to compact, skipping", "level", level)
+		return
+	}
 	logger.Debug("Picked nodes for compaction", "level", level, "node_count", len(pickedNodes))
 
 	// insert to level i + 1 target sstWriter
@@ -187,6 +191,10 @@ func (t *LSMTree) compactMemTable(memCompactItem *memTableCompactItem) {
 }
 
 func (t *LSMTree) pickCompactNodes(level int) []*Node {
+	if level < 0 || level >= len(t.nodes) || len(t.nodes[level]) == 0 {
+		return nil
+	}
+
 	// read half nodes
 	startKey := t.nodes[level][0].Start()
 	endKey := t.nodes[level][0].End()

